File: framework/languages/java/samples/java/ZoneWorld/Server/src/main/java/systems/zlink/samples/zoneworld/server/zone/handlers/BorderSubscriptionHandlers.java
package systems.zlink.samples.zoneworld.server.zone.handlers;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;
import systems.zlink.framework.handlers.ZLinkHandlerGroup;
import systems.zlink.framework.handlers.ZLinkSpotSubscription;
import systems.zlink.samples.zoneworld.server.zone.spots.ZoneSpot;
import systems.zlink.samples.zoneworld.shared.Messages;
import systems.zlink.samples.zoneworld.shared.ZoneWorldNames;
@ZLinkHandlerGroup(ZoneWorldNames.ZONE_CHANNEL)
public final class BorderSubscriptionHandlers {
    private BorderSubscriptionHandlers() {
    }

    // --8<-- [start:doc-zw-border-subscribe]
    public static Class<?> forRoute(String fromZoneId, String toZoneId) {
        return switch (ZoneWorldNames.borderTopic(fromZoneId, toZoneId)) {
            case ZoneWorldNames.NW_NE -> NorthWestToNorthEast.class;
            case ZoneWorldNames.NW_SW -> NorthWestToSouthWest.class;
            case ZoneWorldNames.NE_NW -> NorthEastToNorthWest.class;
            case ZoneWorldNames.NE_SE -> NorthEastToSouthEast.class;
            case ZoneWorldNames.SW_NW -> SouthWestToNorthWest.class;
            case ZoneWorldNames.SW_SE -> SouthWestToSouthEast.class;
            case ZoneWorldNames.SE_NE -> SouthEastToNorthEast.class;
            case ZoneWorldNames.SE_SW -> SouthEastToSouthWest.class;
            default -> throw new IllegalArgumentException("unknown border route: "
                + fromZoneId + " -> " + toZoneId);
        };
    }
    // --8<-- [end:doc-zw-border-subscribe]

    private static CompletionStage<Void> apply(ZoneSpot spot, Messages.ZoneBorderEvent event) {
        spot.applyBorder(event);
        return CompletableFuture.completedFuture(null);
    }

    @ZLinkSpotSubscription(topic = ZoneWorldNames.NW_NE)
    public static final class NorthWestToNorthEast {
        public CompletionStage<Void> handle(ZoneSpot spot, Messages.ZoneBorderEvent event) { return apply(spot, event); }
    }
    @ZLinkSpotSubscription(topic = ZoneWorldNames.NW_SW)
    public static final class NorthWestToSouthWest {
        public CompletionStage<Void> handle(ZoneSpot spot, Messages.ZoneBorderEvent event) { return apply(spot, event); }
    }
    @ZLinkSpotSubscription(topic = ZoneWorldNames.NE_NW)
    public static final class NorthEastToNorthWest {
        public CompletionStage<Void> handle(ZoneSpot spot, Messages.ZoneBorderEvent event) { return apply(spot, event); }
    }
    @ZLinkSpotSubscription(topic = ZoneWorldNames.NE_SE)
    public static final class NorthEastToSouthEast {
        public CompletionStage<Void> handle(ZoneSpot spot, Messages.ZoneBorderEvent event) { return apply(spot, event); }
    }
    @ZLinkSpotSubscription(topic = ZoneWorldNames.SW_NW)
    public static final class SouthWestToNorthWest {
        public CompletionStage<Void> handle(ZoneSpot spot, Messages.ZoneBorderEvent event) { return apply(spot, event); }
    }
    @ZLinkSpotSubscription(topic = ZoneWorldNames.SW_SE)
    public static final class SouthWestToSouthEast {
        public CompletionStage<Void> handle(ZoneSpot spot, Messages.ZoneBorderEvent event) { return apply(spot, event); }
    }
    @ZLinkSpotSubscription(topic = ZoneWorldNames.SE_NE)
    public static final class SouthEastToNorthEast {
        public CompletionStage<Void> handle(ZoneSpot spot, Messages.ZoneBorderEvent event) { return apply(spot, event); }
    }
    @ZLinkSpotSubscription(topic = ZoneWorldNames.SE_SW)
    public static final class SouthEastToSouthWest {
        public CompletionStage<Void> handle(ZoneSpot spot, Messages.ZoneBorderEvent event) { return apply(spot, event); }
    }
}

File: framework/languages/java/samples/kotlin/ZoneWorld/Server/src/main/kotlin/systems/zlink/samples/kotlin/zoneworld/dynamic/BorderSubscriptionHandlers.kt
package systems.zlink.samples.kotlin.zoneworld.dynamic

import systems.zlink.framework.handlers.ZLinkSpotSubscription
import systems.zlink.framework.kotlin.ZLinkSuspendingSpotSubscriptionHandler
import systems.zlink.framework.kotlin.addHandler
import systems.zlink.framework.spots.ZLinkSpotHandlerRegistry
import systems.zlink.samples.kotlin.zoneworld.server.zone.ZoneSpot
import systems.zlink.samples.kotlin.zoneworld.shared.Messages
import systems.zlink.samples.kotlin.zoneworld.shared.ZoneWorldNames

class BorderSubscriptionHandlers {
    companion object {
        fun registerForRoute(
            handlers: ZLinkSpotHandlerRegistry,
            fromZoneId: String,
            toZoneId: String,
        ) {
            when (ZoneWorldNames.borderTopic(fromZoneId, toZoneId)) {
                ZoneWorldNames.NW_NE -> handlers.addHandler<NorthWestToNorthEast>()
                ZoneWorldNames.NW_SW -> handlers.addHandler<NorthWestToSouthWest>()
                ZoneWorldNames.NE_NW -> handlers.addHandler<NorthEastToNorthWest>()
                ZoneWorldNames.NE_SE -> handlers.addHandler<NorthEastToSouthEast>()
                ZoneWorldNames.SW_NW -> handlers.addHandler<SouthWestToNorthWest>()
                ZoneWorldNames.SW_SE -> handlers.addHandler<SouthWestToSouthEast>()
                ZoneWorldNames.SE_NE -> handlers.addHandler<SouthEastToNorthEast>()
                ZoneWorldNames.SE_SW -> handlers.addHandler<SouthEastToSouthWest>()
                else -> error("unknown border route: $fromZoneId -> $toZoneId")
            }
        }

        suspend fun apply(spot: ZoneSpot, event: Messages.ZoneBorderEvent) {
            spot.applyBorder(event)
        }
    }

    @ZLinkSpotSubscription(topic = ZoneWorldNames.NW_NE)
    class NorthWestToNorthEast :
        ZLinkSuspendingSpotSubscriptionHandler<ZoneSpot, Messages.ZoneBorderEvent> {
        override suspend fun handle(spot: ZoneSpot, event: Messages.ZoneBorderEvent) =
            apply(spot, event)
    }

    @ZLinkSpotSubscription(topic = ZoneWorldNames.NW_SW)
    class NorthWestToSouthWest :
        ZLinkSuspendingSpotSubscriptionHandler<ZoneSpot, Messages.ZoneBorderEvent> {
        override suspend fun handle(spot: ZoneSpot, event: Messages.ZoneBorderEvent) =
            apply(spot, event)
    }

    @ZLinkSpotSubscription(topic = ZoneWorldNames.NE_NW)
    class NorthEastToNorthWest :
        ZLinkSuspendingSpotSubscriptionHandler<ZoneSpot, Messages.ZoneBorderEvent> {
        override suspend fun handle(spot: ZoneSpot, event: Messages.ZoneBorderEvent) =
            apply(spot, event)
    }

    @ZLinkSpotSubscription(topic = ZoneWorldNames.NE_SE)
    class NorthEastToSouthEast :
        ZLinkSuspendingSpotSubscriptionHandler<ZoneSpot, Messages.ZoneBorderEvent> {
        override suspend fun handle(spot: ZoneSpot, event: Messages.ZoneBorderEvent) =
            apply(spot, event)
    }

    @ZLinkSpotSubscription(topic = ZoneWorldNames.SW_NW)
    class SouthWestToNorthWest :
        ZLinkSuspendingSpotSubscriptionHandler<ZoneSpot, Messages.ZoneBorderEvent> {
        override suspend fun handle(spot: ZoneSpot, event: Messages.ZoneBorderEvent) =
            apply(spot, event)
    }

    @ZLinkSpotSubscription(topic = ZoneWorldNames.SW_SE)
    class SouthWestToSouthEast :
        ZLinkSuspendingSpotSubscriptionHandler<ZoneSpot, Messages.ZoneBorderEvent> {
        override suspend fun handle(spot: ZoneSpot, event: Messages.ZoneBorderEvent) =
            apply(spot, event)
    }

    @ZLinkSpotSubscription(topic = ZoneWorldNames.SE_NE)
    class SouthEastToNorthEast :
        ZLinkSuspendingSpotSubscriptionHandler<ZoneSpot, Messages.ZoneBorderEvent> {
        override suspend fun handle(spot: ZoneSpot, event: Messages.ZoneBorderEvent) =
            apply(spot, event)
    }

    @ZLinkSpotSubscription(topic = ZoneWorldNames.SE_SW)
    class SouthEastToSouthWest :
        ZLinkSuspendingSpotSubscriptionHandler<ZoneSpot, Messages.ZoneBorderEvent> {
        override suspend fun handle(spot: ZoneSpot, event: Messages.ZoneBorderEvent) =
            apply(spot, event)
    }
}
